ga4gh/backend.py

```diff
@@ -762,6 +770,69 @@ def runSearchDatasets(self, request):
             protocol.SearchDatasetsResponse,
             self.datasetsGenerator)
 
+    def runSearchGenotypePhenotype(self, requestStr):
+        """
+        Runs the specified SearchGenotypePhenotypeRequest.
+        We do not attempt to re-use runSearchRequest as
+        All you really need is to get the offset from the pageToken, and then
+        pass this on to the query method in the data model object.
+        """
+        # TODO verify evidence maps to drug
+
+        # validate request
+        self.startProfile()
+        try:
+            requestDict = json.loads(requestStr)
+        except ValueError:
+            raise exceptions.InvalidJsonException(requestStr)
+        requestClass = protocol.SearchGenotypePhenotypeRequest
+        responseClass = protocol.SearchGenotypePhenotypeResponse
+        self.validateRequest(requestDict, requestClass)
+        request = requestClass.fromJsonDict(requestDict)
+        if request.pageSize is None:
+            request.pageSize = self._defaultPageSize
+        if request.pageSize <= 0:
+            raise exceptions.BadPageSizeException(request.pageSize)
+        responseBuilder = protocol.SearchResponseBuilder(
+            responseClass, request.pageSize, self._maxResponseLength)
+        # ensure request has data we can query
+        if (request.evidence is None and
+                request.phenotype is None and
+                request.feature is None):
+            msg = "Error:One of evidence,phenotype or feature must be non-null"
+            raise exceptions.BadRequestException(msg)
+        # determine offset for paging
+        if request.pageToken is not None:
+            request.offset = int(request.pageToken.split(':')[1])
+        else:
+            request.offset = 0
+
+        offset = request.offset
+        count = 0
+        iterator = self._g2pDataset._search(request)
+        for obj in iterator:
+            responseBuilder.addValue(obj)
+            offset += 1
+            count += 1
+            if responseBuilder.isFull():
+                break
+
+        nextPageToken = None
+        if request.pageSize == count:
+            nextPageToken = "{}:{}".format(0, offset)
+
+        responseBuilder.setNextPageToken(nextPageToken)
+        responseString = responseBuilder.getJsonString()
+        self.validateResponse(responseString, responseClass)
+        self.endProfile()
+        return responseString
+
+        # # old ......
+        # return self.runSearchRequest(
+        #     request, protocol.SearchGenotypePhenotypeRequest,
+        #     protocol.SearchGenotypePhenotypeResponse,
+        #     self.genotypePhenotypeGenerator)
+
 
```

Comment:
Thanks.  I've added this change.  As you say, we will determine how much of a problem it will be in practice.    One future change in backend.py might be to separate out "total number of rows"  from "length of the collection in the request".